vm/state_migration/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+//! Common code that's shared across all migration code.
+//! Each network upgrade / state migration code lives in their own module.
+
+use address::Address;
+use cid::Cid;
+use clock::ChainEpoch;
+use ipld_blockstore::BlockStore;
+use std::error::Error as StdError;
+use std::rc::Rc;
+use vm::{ActorState, TokenAmount};
+
+pub mod nv12;
+
+pub type MigrationResult<T> = Result<T, MigrationError>;
+
+#[derive(thiserror::Error, Debug)]
+pub enum MigrationError {
+    #[error("Failed creating job for state migration")]
+    MigrationJobCreate(Box<dyn StdError>),
```

Comment:
Ah yeah, i agree. Forgot that we agreed that new code we do should try out best not to use dyn errors.

Reply:
Yeah, i do feel uneasy with using Box<dyn Error>, but i added them because not adding them is not propagating the underlying error source, thereby providing less context to end user.

Reply:
Just wrap the underlying error type: https://github.com/ChainSafe/forest/blob/main/blockchain/chain_sync/src/chain_muxer.rs#L57

That way the caller can also handle different error types separately when needed.

Reply:
Yes, could have easily done that. But, the underlying error type itself is erased and is also `Box<dyn Error>`

Reply:
for example: for_each method of StateTree returns an erased error type:

```rust
    pub fn for_each<F>(&self, mut f: F) -> Result<(), Box<dyn StdError>>
    where
        F: FnMut(Address, &ActorState) -> Result<(), Box<dyn StdError>>,
        S: BlockStore,
    {
        self.hamt.for_each(|k, v| f(Address::from_bytes(&k.0)?, v))
    }
```

Reply:
I see. Seems like we need to address this in our core libraries before it makes sense to enforce across the board.

Reply:
Makes sense. Everything that implements the Error trait i think has `.to_string()`. Perhaps we can map all Box<dyn Error> to Strings until we change the underlying libraries to return Error types? Does that sound reasonable?

Reply:
Yeah, can do that for the time being